segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/fs/FSStorage.java

```diff
@@ -0,0 +1,385 @@
+ */
+package io.pravega.segmentstore.storage.impl.fs;
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.util.ImmutableDate;
+import io.pravega.segmentstore.contracts.BadOffsetException;
+import io.pravega.segmentstore.contracts.SegmentProperties;
+import io.pravega.segmentstore.contracts.StreamSegmentExistsException;
+import io.pravega.segmentstore.contracts.StreamSegmentInformation;
+import io.pravega.segmentstore.contracts.StreamSegmentNotExistsException;
+import io.pravega.segmentstore.contracts.StreamSegmentSealedException;
+import io.pravega.segmentstore.storage.SegmentHandle;
+import io.pravega.segmentstore.storage.Storage;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.RandomAccessFile;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.FileChannel;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.file.AccessDeniedException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.nio.file.attribute.FileAttribute;
+import java.nio.file.attribute.PosixFileAttributes;
+import java.nio.file.attribute.PosixFilePermission;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+
+import static java.nio.file.attribute.PosixFilePermission.OWNER_WRITE;
+
+/**
+ * Storage adapter for file system based Tier 2.
+ *
+ * Each segment is represented as a single file on the underlying storage. As the data in Tier 2 is not modified
+ * once written, any attempt to re-write data with the same file offset does not cause any form of inconsistency
+ * as the bytes are the same and they are in the same position.
+ */
+
+@Slf4j
+public class FSStorage implements Storage {
+
+    //region members
+
+    private final FSStorageConfig config;
+    private final ExecutorService executor;
+
+    //endregion
+
+    //region constructor
+
+    /**
+     * Creates a new instance of the FSStorage class.
+     *
+     * @param config   The configuration to use.
+     * @param executor The executor to use for running async operations.
+     */
+    public FSStorage(FSStorageConfig config, ExecutorService executor) {
+        Preconditions.checkNotNull(config, "config");
+        Preconditions.checkNotNull(executor, "executor");
+        this.config = config;
+        this.executor = executor;
+    }
+
+    //endregion
+
+    //region Storage implementation
+    @Override
+    public void initialize(long containerEpoch) {
+
```

Comment:
Epoc was introduced to enable distributed locking for HDFS. As we do not need locking (because of write at an offset provided by FS), the initialize function is no op.